src/svm/svr.rs

```diff
@@ -0,0 +1,538 @@
+            }
+            if g > self.gmax && a < self.c {
+                self.gmax = g;
+                self.gmaxindex = 0;
+                self.svmax = i;
+            }
+
+            let g = v.grad[1];
+            let a = v.alpha[1];
+            if g < self.gmin && a < self.c {
+                self.gmin = g;
+                self.gminindex = 1;
+                self.svmin = i;
+            }
+            if g > self.gmax && a > T::zero() {
+                self.gmax = g;
+                self.gmaxindex = 1;
+                self.svmax = i;
+            }
+        }
+    }
+
+    fn smo(mut self) -> (Vec<M::RowVector>, Vec<T>, T) {
+        let cache: Cache<T> = Cache::new(self.sv.len());
+
+        self.minmax();
+
+        while self.gmax - self.gmin > self.tol {
+            let v1 = self.svmax;
+            let i = self.gmaxindex;
+            let old_alpha_i = self.sv[v1].alpha[i];
+
+            let k1 = cache.get(self.sv[v1].index, || {
+                self.sv
+                    .iter()
+                    .map(|vi| self.kernel.apply(&self.sv[v1].x, &vi.x))
+                    .collect()
+            });
+
+            let mut v2 = self.svmin;
+            let mut j = self.gminindex;
+            let mut old_alpha_j = self.sv[v2].alpha[j];
+
+            let mut best = T::zero();
+            let gi = if i == 0 {
+                -self.sv[v1].grad[0]
+            } else {
+                self.sv[v1].grad[1]
+            };
+            for jj in 0..self.sv.len() {
+                let v = &self.sv[jj];
+                let mut curv = self.sv[v1].k + v.k - T::two() * k1[v.index];
+                if curv <= T::zero() {
+                    curv = self.tau;
+                }
+
+                let mut gj = -v.grad[0];
+                if v.alpha[0] > T::zero() && gj < gi {
+                    let gain = -((gi - gj) * (gi - gj)) / curv;
+                    if gain < best {
+                        best = gain;
+                        v2 = jj;
+                        j = 0;
+                        old_alpha_j = self.sv[v2].alpha[0];
+                    }
+                }
+
+                gj = v.grad[1];
+                if v.alpha[1] < self.c && gj < gi {
+                    let gain = -((gi - gj) * (gi - gj)) / curv;
+                    if gain < best {
+                        best = gain;
+                        v2 = jj;
+                        j = 1;
+                        old_alpha_j = self.sv[v2].alpha[1];
+                    }
+                }
+            }
+
+            let k2 = cache.get(self.sv[v2].index, || {
+                self.sv
+                    .iter()
+                    .map(|vi| self.kernel.apply(&self.sv[v2].x, &vi.x))
+                    .collect()
+            });
+
+            let mut curv = self.sv[v1].k + self.sv[v2].k - T::two() * k1[self.sv[v2].index];
+            if curv <= T::zero() {
+                curv = self.tau;
+            }
+
+            if i != j {
```

Comment:
I dont know if it is applicable in this particular case, but I believe that the Rust way of doing this is patter matching (`match` `case`)

Reply:
Usually I use `match` when I have 2 or more conditions to go through. In this case I have a simple if/else case that doesn't justify matching.